### PR TITLE
LF: hashing for committed contracts

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
, i.e. contracts that may containt resolve contract ids.

CHANGELOG_BEGIN
CHANGLEOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 1]
```diff
@@ -319,6 +321,15 @@ object Hash {
   ): Either[String, Hash] =
     handleError(assertHashContractKey(templateId, key))
 
+  def hashCommittedContract(
+      templateId: Ref.Identifier,
+      key: Value[Value.ContractId],
```
Comment (oliverse-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala:326]: key should be arg actually, right (`s/key/arg/) ? Also do we not need to include all `ContractInst` parameters, i.e. `agreementText` too?

More on the stylistic side on which I am flexible, just my 2 cents: I was assuming we'd actually pass in a `ContractInst` and not necessarily refer to the contract as committed.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala:326]: You right, `"key"` should be `"arg"`.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala]: I am more incline to ignore the `agreementText` as it is not semantically part of the contract and we may want to drop it later.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala]: I add a function for `Value.ContractInst[VersionedValue[ContractId]]`

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala [hunk 2]
```diff
@@ -339,11 +339,13 @@ object Value extends CidContainer1[Value] {
 
   sealed abstract class ContractId extends Product with Serializable {
     def coid: String
+    def toBytes: Bytes
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala:342]: You don't seem to make use of the method in this PR. Is this an unrelated change?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala [hunk 3]
```diff
@@ -339,11 +339,13 @@ object Value extends CidContainer1[Value] {
...
   }
 
   object ContractId {
     final case class V0(coid: Ref.ContractIdString) extends ContractId {
       override def toString: String = s"ContractId($coid)"
+      lazy val toBytes: Bytes = V0.prefix ++ Utf8.getBytes(coid)
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala:348]: I'm surprised that we need another prefix. Overall, I find this API a bit confusing:
* `Ref.ContractIdString` already requires that V0 contract IDs start with a `#`. So we now have two mandatory prefixes here. Isn't `#` enough different from a `00` prefix? Do you want `toBytes` to always return a valid Base16 encoding?
* `V1`'s `toString` method includes the prefix `00`, but `V0` does not.
* The order for contract IDs decrees that `V0` comes before `V1`, but the prefixes are the opposite (`ff` vs `00`), even though the reference string itself is ordered lexicographically.

Can we make this whole thing simpler and more consistent rather than adding more on top of it?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala (hunk 1 above)
Comment (andreaslochbihler-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala]: Re agreement text: I always thought that the agreement text is just a DAML function of the template arguments. If that's also the case in DAML-LF, then no need to add the agreement text.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala]: It wa ;). V0 and V1 override this.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala]: You are complexity right. 
I simplify by supporting only V1 in the hashing. 
Only Sandbox classic supports V0. We will see how to deal with that later, if we need the hashing in sandbox.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 4]
```diff
@@ -313,12 +315,38 @@ object Hash {
   def safeHashContractKey(templateId: Ref.Identifier, key: Value[Nothing]): Hash =
     assertHashContractKey(templateId, key)
 
-  def hashContractKey(
+  private def comittedCidToBytes(coid: ContractId): Bytes =
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala:318]:
```suggestion
  private def committedCidToBytes(coid: ContractId): Bytes =
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala (hunk 2 above)
Comment (andreaslochbihler-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala]: Sorry for the confusion. I missed the `_.bytes` in `hashCommittedContract`. 🤦

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 5]
```diff
@@ -313,12 +315,38 @@ object Hash {
...
+    coid match {
+      case coid: ContractId.V1 => coid.toBytes
+      case _: ContractId.V0 => error("hashing of Contract ID V0 not supported")
+    }
+
+  // This function assumes that key is well typed, i.e. :
+  // 1 - `templateId` is the identifier for a template with a key of type τ
+  // 2 - `key` is a value of type τ
+  @throws[HashingError]
+  def assertHashContractKey(
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala:328]: Minor: Given that this method returns an `Either`, I'm somewhat surprised that it has `assert` in its name.

**Review by andreaslochbihler-da — APPROVED** [review]

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 6]
```diff
@@ -319,6 +321,29 @@ object Hash {
   ): Either[String, Hash] =
     handleError(assertHashContractKey(templateId, key))
 
+  private[this] def committedCidToBytes(coid: ContractId): Bytes =
+    coid match {
+      case coid: ContractId.V1 => coid.toBytes
```
Comment (oggy-) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala:326]: As I mentioned earlier, I'm in the process of figuring out what happens if we limit this to `coid.discriminator.bytes`. I want to put out a short doc about this by the end of the week. I don't mind merging this PR as-is though, as this hashing functionality won't be used elsewhere yet.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala:326]: I switched back the PR in draft mode while waiting for your conclusions

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala (hunk 1 above)
Comment (cocreature) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala]: That’s correct.
